edit pass: Grpc batch one (#15898)
dotnet/docs: docs/architecture/grpc-for-wcf-developers/kubernetes.md

---
title: Kubernetes - gRPC for WCF developers
description: Running ASP.NET Core gRPC services in a Kubernetes cluster.
ms.date: 09/02/2019
---

# Kubernetes

> Although it's possible to run containers manually on Docker hosts, for reliable production systems it's better to use a container orchestration engine to manage multiple instances running across several servers in a cluster. There are various container orchestration engines available, including Kubernetes, Docker Swarm, and Apache Mesos. But of these engines, Kubernetes is far and away the most widely used, so it will be the focus of this chapter.

Reviewer: I'm really not a fan of the term `container orchestration engines`. I think container orchestrator is a better fit here. Engine seems oddly specific

Reviewer: I'm okay not making a change here.

  # Object properties
```

> The `apiVersion` property is used to specify which version (and which API) the file is intended for. The `kind` property specifies the kind of object the YAML represents. The `metadata` property contains object properties like `name`, `namespace`, and `labels`.

Reviewer: This phrasing is a little confusing since `name` is mandatory property on `metadata`, but namespace is optional. Both AND and OR seem slightly misleading to me.

Author: To me, "and" makes more sense than "or" because we're just talking about what the property contains. Do you want to add clarity about which are optional and which are required?

Reviewer: Let's not add complexity here. I was being extremely pedantic here, but on second reading I'm fine with `and`.

Most Kubernetes YAML files will also have a `spec` section that describes the resources and configuration necessary to create the object.

### Pods

Pods are the basic units of execution in Kubernetes. They can run multiple containers, but they're also used to run single containers. The pod also includes any storage resources required by the containers, and the network IP address.

### Services

Services are meta-objects that describe Pods (or sets of Pods) and provide a way to access them within the cluster, such as mapping a service name to a set of pod IP addresses by using the cluster DNS service.

### Deployments

> Deployments are the *described state* objects for Pods. If you create a pod manually, it won't be restarted when it terminates. Deployments are used to tell the cluster which Pods, and how many replicas of those Pods, should be running at the present time.

Reviewer: ```suggestion
Deployments are the *desired state* objects for Pods. If you create a pod manually, it won't be restarted when it terminates. Deployments are used to tell the cluster which Pods, and how many replicas of those Pods, should be running at the present time.
```

### Other objects

Pods, Services, and Deployments are just three of the most basic object types. There are dozens of other object types that are managed by Kubernetes clusters. For more information, see the [Kubernetes Concepts](https://kubernetes.io/docs/concepts/) documentation.

### Namespaces

Kubernetes clusters are designed to scale to hundreds or thousands of nodes and to run similar numbers of services. To avoid clashes between object names, namespaces are used to group objects together as part of larger applications. Kubernetes's own services run in a `default` namespace. All user objects should be created in their own namespaces to avoid potential clashes with default objects or other tenants in the cluster.

## Get started with Kubernetes

> If you're running Docker Desktop for Windows and macOS, Kubernetes is already available. Just enable it in the **Kubernetes** section of the **Settings** window:

Reviewer: ```suggestion
If you're running Docker Desktop for Windows or Docker Desktop for Mac, Kubernetes is already available. Just enable it in the **Kubernetes** section of the **Settings** window:
```

The word `and` changed the meaning here. To make it explicit I used the verbatim names of the products
